// image-metadata-editor/dist/commonjs/index.js
"use strict";
/*
 * run with:

npx tsc -p tsconfig.json   # Generate ES6 modules
npx tsc -p tsconfig.commonjs.json  # Generate CommonJS modules

 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.readWebpXMP = exports.readXMP = exports.readPNGtext = exports.getMimeType = exports.writeXMP = exports.writeExif = exports.writeWebpXMP = exports.writeJPGMarker = exports.writePNGtext = void 0;
const crc_32_1 = __importDefault(require("crc-32"));
const { buf } = crc_32_1.default;
const fast_xml_parser_1 = __importDefault(require("fast-xml-parser"));
const pako_1 = __importDefault(require("pako"));
const utils_js_1 = require("./utils.js");
/**
 * Writes tEXt metadata to a PNG Uint8Array.
 * @param {Blob|Uint8Array} image - The original PNG Uint8Array.
 * @param {string} key - The metadata key.
 * @param {string} value - The metadata value.
 * @param {string} mode - The encoding mode to use. Default = "tEXt".
 * @returns {Promise<Uint8Array>} - Updated Uint8Array.
 *
 * @throws {TypeError} If a non PNG image is used
 * @throws {TypeError} If using an unsupported text mode
 */
function writePNGtext(image, key, value, mode = "tEXt") {
    return __awaiter(this, void 0, void 0, function* () {
        (0, utils_js_1.typeCheck)(key, ["string"]);
        (0, utils_js_1.typeCheck)(value, ["string"]);
        (0, utils_js_1.typeCheck)(mode, ["string"]);
        const imageArr = yield (0, utils_js_1.cast2Uint8Array)(image);
        if ((yield getMimeType(imageArr)) !== "image/png") {
            throw new TypeError("A PNG is required to use writePNGtEXt");
        }
        if (!["tEXt", "zTXt"].includes(mode)) {
            throw new TypeError("Unsupported text mode");
        }
        var keyValue = `${key}\0${value}`;
        var keyValueBytes = new TextEncoder().encode(keyValue);
        if (mode === "zTXt") {
            const keyBytes = new TextEncoder().encode(`${key}\0\0`);
            const ValueBytes = pako_1.default.deflate(value);
            var keyValueBytes = new Uint8Array(keyBytes.length + ValueBytes.length);
            keyValueBytes.set(keyBytes);
            keyValueBytes.set(ValueBytes, keyBytes.length);
        }
        const tEXtKeyValue = new Uint8Array(keyValueBytes.length + 4);
        tEXtKeyValue.set(new TextEncoder().encode(mode));
        tEXtKeyValue.set(keyValueBytes, 4);
        const crc32 = buf(tEXtKeyValue);
        const keyValueLen = new Uint8Array(4);
        new DataView(keyValueLen.buffer).setUint32(0, keyValueBytes.length, false);
        // Build the chunk_tEXt
        const tEXt = new Uint8Array(keyValueBytes.length + 12);
        tEXt.set(keyValueLen, 0);
        tEXt.set(tEXtKeyValue, 4);
        new DataView(tEXt.buffer).setUint32(keyValueBytes.length + 8, crc32, false);
        // Update the PNG with the chunk_tEXt
        const insertPosition = new TextDecoder().decode(imageArr).indexOf("IHDR") + 21;
        const imgBytesBefore = imageArr.slice(0, insertPosition);
        const imgBytesAfter = imageArr.slice(insertPosition);
        const newPng = new Uint8Array(imgBytesBefore.length + tEXt.length + imgBytesAfter.length);
        newPng.set(imgBytesBefore);
        newPng.set(tEXt, imgBytesBefore.length);
        newPng.set(imgBytesAfter, imgBytesBefore.length + tEXt.length);
        return newPng;
    });
}
exports.writePNGtext = writePNGtext;
/**
 * Writes metadata to a JPG Uint8Array.
 * @param {Blob|Uint8Array} image - The original JPG Uint8Array.
 * @param {string} value - The metadata value.
 * @param {number[]} marker - The jpg marker, defaults to COM.
 * @returns {Promise<Uint8Array>} - Updated Uint8Array.
 *
 * @throws {TypeError} If a non JPG image is used
 * @throws {TypeError} If a marker is invalid
 */
function writeJPGMarker(image, value, marker = [0xff, 0xfe]) {
    return __awaiter(this, void 0, void 0, function* () {
        (0, utils_js_1.typeCheck)(value, ["string"]);
        (0, utils_js_1.typeCheck)(marker, [(Array)]);
        const imageArr = yield (0, utils_js_1.cast2Uint8Array)(image);
        if ((yield getMimeType(imageArr)) !== "image/jpeg") {
            throw new TypeError("A JPG is required to use writeJPGCOM");
        }
        if (marker[0] !== 0xff) {
            throw new TypeError("Invalid Marker");
        }
        const valueBytes = new TextEncoder().encode(value);
        const tEXt = new Uint8Array(4 + valueBytes.length);
        tEXt.set(marker, 0);
        new DataView(tEXt.buffer).setUint16(2, valueBytes.length + 2, false);
        tEXt.set(valueBytes, 4);
        // Update the JPG
        const insertPosition = 2;
        const imgBytesBefore = imageArr.slice(0, insertPosition);
        const imgBytesAfter = imageArr.slice(insertPosition);
        const newJpg = new Uint8Array(imgBytesBefore.length + tEXt.length + imgBytesAfter.length);
        newJpg.set(imgBytesBefore);
        newJpg.set(tEXt, imgBytesBefore.length);
        newJpg.set(imgBytesAfter, imgBytesBefore.length + tEXt.length);
        return newJpg;
    });
}
exports.writeJPGMarker = writeJPGMarker;
/**
 * Writes metadata to a Webp Uint8Array.
 * @param {Blob|Uint8Array} image - The original Webp Uint8Array.
 * @param {string} value - The metadata value.
 * @returns {Promise<Uint8Array>} - Updated Uint8Array.
 *
 * @throws {TypeError} If a non Webp image is used
 * @throws {TypeError} If a marker is invalid
 */
function writeWebpXMP(image, value) {
    return __awaiter(this, void 0, void 0, function* () {
        (0, utils_js_1.typeCheck)(value, ["string"]);
        const imageArr = yield (0, utils_js_1.cast2Uint8Array)(image);
        if ((yield getMimeType(imageArr)) !== "image/webp") {
            throw new TypeError("A webp is required to use writeWebpXMP");
        }
        throw new TypeError("NotImplemented :(");
    });
}
exports.writeWebpXMP = writeWebpXMP;
/**
 * Writes EXIF metadata to an image Uint8Array.
 * @param {Blob|Uint8Array} image - The original image Uint8Array.
 * @param {string} key - The EXIF key.
 * @param {string} value - The EXIF value.
 * @returns {Promise<Uint8Array>} - Updated Uint8Array.
 *
 * @throws {TypeError} NotImplemented
 */
function writeExif(image, key, value) {
    return __awaiter(this, void 0, void 0, function* () {
        (0, utils_js_1.typeCheck)(key, ["string"]);
        (0, utils_js_1.typeCheck)(value, ["string"]);
        throw new TypeError("NotImplemented :(");
    });
}
exports.writeExif = writeExif;
/**
 * Writes XMP metadata to an image Uint8Array.
 * @param {Blob|Uint8Array} image - The original image Uint8Array.
 * @param {string} value - The XMP value.
 * @returns {Promise<Uint8Array>} - Updated Uint8Array.
 *
 * @throws {TypeError} If the XMP is invalid/ badly structured
 * @throws {TypeError} If an unsupported image is used
 */
function writeXMP(image, value) {
    return __awaiter(this, void 0, void 0, function* () {
        (0, utils_js_1.typeCheck)(value, ["string"]);
        const imageArr = yield (0, utils_js_1.cast2Uint8Array)(image);
        const err = fast_xml_parser_1.default.XMLValidator.validate(value);
        if (err !== true) {
            throw new TypeError("XMP must be valid XML...\n" + err.err.code + ": " + err.err.msg);
        }
        if ((yield getMimeType(imageArr)) === "image/png") {
            return writePNGtext(imageArr, "XML:com.adobe.xmp", value);
        }
        if ((yield getMimeType(imageArr)) === "image/jpeg") {
            return writeJPGMarker(imageArr, `http://ns.adobe.com/xap/1.0/\0${value}`, [0xff, 0xe1]);
        }
        if ((yield getMimeType(imageArr)) === "image/webp") {
            return writeWebpXMP(imageArr, value);
        }
        throw new TypeError("This image is not supported :(");
    });
}
exports.writeXMP = writeXMP;
/**
 * Get the MIME type of an image based on its magic bytes.
 * @param {Blob|Uint8Array} image - The image as a Uint8Array.
 * @returns {string} - The MIME type (e.g., "image/png", "image/jpeg", "image/webp"), or "null" if the format is not recognized.
 */
function getMimeType(image) {
    return __awaiter(this, void 0, void 0, function* () {
        const imageArr = yield (0, utils_js_1.cast2Uint8Array)(image);
        const magic = imageArr.slice(0, 4);
        if (magic[0] === 0x89 &&
            magic[1] === 0x50 &&
            magic[2] === 0x4e &&
            magic[3] === 0x47) {
            return "image/png";
        }
        else if (magic[0] === 0xff && magic[1] === 0xd8 && magic[2] === 0xff) {
            return "image/jpeg";
        }
        else if (magic[0] === 0x52 &&
            magic[1] === 0x49 &&
            magic[2] === 0x46 &&
            magic[3] === 0x46) {
            // Check for the WebP RIFF header
            return "image/webp";
        }
        else {
            return "null"; // Unknown format
        }
    });
}
exports.getMimeType = getMimeType;
/**
 * Reads tEXt metadata to a PNG Uint8Array.
 * @param {Blob|Uint8Array} image - The original PNG Uint8Array.
 * @param {string} key - The metadata key.
 * @returns {string} - The metadata value.
 *
 * @throws {TypeError} If a non PNG image is used
 * @throws {TypeError} If the value cannot be found
 */
function readPNGtext(image, key) {
    return __awaiter(this, void 0, void 0, function* () {
        (0, utils_js_1.typeCheck)(key, ["string"]);
        const imageArr = yield (0, utils_js_1.cast2Uint8Array)(image);
        if ((yield getMimeType(imageArr)) !== "image/png") {
            throw new TypeError("A PNG is required to use writePNGtEXt");
        }
        var readPosition = 0;
        var mode = "";
        for (mode of ["tEXt", "zTXt"]) {
            readPosition =
                new TextDecoder().decode(imageArr).indexOf(`${mode}${key}`) - 3;
            if (readPosition !== -4) {
                break;
            }
        }
        if (readPosition === -4) {
            throw new TypeError(`"${key}" not found!`);
        }
        const keyValueLen = imageArr.slice(readPosition, readPosition + 4);
        const size = keyValueLen[3] +
            (keyValueLen[2] << 8) +
            (keyValueLen[1] << 16) +
            (keyValueLen[0] << 24);
        const keyValueOffset = readPosition + 8;
        const keyValueBytes = imageArr.slice(keyValueOffset, keyValueOffset + size);
        if (mode === "zTXt") {
            const ValueBytes = keyValueBytes.slice(key.length + 2, keyValueBytes.length);
            return new TextDecoder().decode(pako_1.default.inflate(ValueBytes));
        }
        const ValueBytes = keyValueBytes.slice(key.length + 1, keyValueBytes.length);
        return new TextDecoder().decode(ValueBytes);
    });
}
exports.readPNGtext = readPNGtext;
/**
 * Reads XMP metadata to an image Uint8Array.
 * @param {Blob|Uint8Array} image - The original image Uint8Array.
 * @returns {string} - The stringified XMP
 *
 * @throws {TypeError} If the XMP is invalid/ badly structured
 * @throws {TypeError} If an unsupported image is used
 */
function readXMP(image) {
    return __awaiter(this, void 0, void 0, function* () {
        const imageArr = yield (0, utils_js_1.cast2Uint8Array)(image);
        if ((yield getMimeType(imageArr)) === "image/png") {
            return readPNGtext(imageArr, "XML:com.adobe.xmp");
        }
        if ((yield getMimeType(imageArr)) === "image/webp") {
            return readWebpXMP(imageArr);
        }
        throw new TypeError("This image is not supported :(");
    });
}
exports.readXMP = readXMP;
/**
 * Reads metadata to a Webp Uint8Array.
 * @param {Blob|Uint8Array} image - The original Webp Uint8Array.
 * @returns {string}
 *
 * @throws {TypeError} If a non Webp image is used
 * @throws {TypeError} If a marker is invalid
 */
function readWebpXMP(image) {
    return __awaiter(this, void 0, void 0, function* () {
        const imageArr = yield (0, utils_js_1.cast2Uint8Array)(image);
        if ((yield getMimeType(imageArr)) !== "image/webp") {
            throw new TypeError("A webp is required to use readWebpXMP");
        }
        throw new TypeError("NotImplemented :(");
    });
}
exports.readWebpXMP = readWebpXMP;
